File: 15-pointers/main.zig
const std = @import("std");
const print = std.debug.print;

fn increment(ptr: *i32) void {
    ptr.* += 1; // 通过指针修改值
}

fn sum(values: []const i32) i32 {
    var total: i32 = 0;
    for (values) |v| {
        total += v;
    }
    return total;
}

pub fn main() void {
    // 1. 基本指针 *T
    var x: i32 = 42;
    const ptr: *i32 = &x;
    print("x = {}, *ptr = {}\n", .{ x, ptr.* });

    // 通过指针修改值
    ptr.* = 100;
    print("After ptr.* = 100: x = {}\n", .{x});

    // 2. 通过指针传参
    increment(&x);
    print("After increment: x = {}\n", .{x}); // 101

    // 3. const 指针 vs 可变指针
    const y: i32 = 10;
    const const_ptr: *const i32 = &y;
    print("const_ptr.* = {}\n", .{const_ptr.*});
    // const_ptr.* = 20; // 编译错误！不能通过 const 指针修改

    // 4. 可选指针 ?*T（替代 null 指针）
    var val: i32 = 55;
    var opt_ptr: ?*i32 = &val;
    if (opt_ptr) |p| {
        print("opt_ptr has value: {}\n", .{p.*});
    }
    opt_ptr = null;
    if (opt_ptr) |_| {
        print("This won't print\n", .{});
    } else {
        print("opt_ptr is null\n", .{});
    }

    // 5. 多元素指针 [*]T 和指针算术
    var arr = [_]i32{ 10, 20, 30, 40, 50 };
    const many_ptr: [*]i32 = &arr;
    print("many_ptr[0] = {}\n", .{many_ptr[0]});
    print("many_ptr[2] = {}\n", .{many_ptr[2]});

    // 指针算术
    const offset_ptr = many_ptr + 2;
    print("(many_ptr + 2)[0] = {}\n", .{offset_ptr[0]}); // 30

    // 6. 多元素指针转切片（更安全）
    const slice: []i32 = arr[0..5];
    print("slice sum = {}\n", .{sum(slice)});

    // 7. 哨兵终止指针 [*:0]u8（C 字符串兼容）
    const c_str: [*:0]const u8 = "Hello from Zig";
    // 转换为切片使用
    const zig_str = std.mem.span(c_str);
    print("C string: {s} (len={})\n", .{ zig_str, zig_str.len });

    // 8. @ptrCast - 指针类型转换
    const bytes = [_]u8{ 0x01, 0x00, 0x00, 0x00 };
    const byte_ptr: [*]const u8 = &bytes;
    // 将字节指针视为 u32 指针（注意字节序）
    const int_ptr: *const u32 = @ptrCast(@alignCast(byte_ptr));
    print("Bytes as u32: {}\n", .{int_ptr.*});

    // 9. 指针与数组的关系
    var nums = [_]i32{ 1, 2, 3 };
    const nums_ptr: *[3]i32 = &nums;
    // 数组指针可以自动转为切片
    const nums_slice: []i32 = nums_ptr;
    print("nums_slice[1] = {}\n", .{nums_slice[1]});
}

File: 15-pointers/main.rs
fn increment(val: &mut i32) {
    *val += 1; // 通过可变引用修改值
}

fn sum(values: &[i32]) -> i32 {
    values.iter().sum()
}

fn main() {
    // 1. 不可变引用 &T（共享借用）
    let x = 42;
    let r = &x;
    println!("x = {}, *r = {}", x, *r);
    // Rust 会自动解引用，所以通常不需要写 *r
    println!("r (auto-deref) = {}", r);

    // 2. 可变引用 &mut T（独占借用）
    let mut y = 100;
    {
        let r_mut = &mut y;
        *r_mut += 1;
        println!("After mutation: {}", r_mut);
    } // r_mut 的借用在此结束
    println!("y = {}", y); // 101

    // 3. 通过可变引用传参
    let mut z = 50;
    increment(&mut z);
    println!("After increment: z = {}", z); // 51

    // 4. 借用规则：同一时刻只能有一个可变引用 OR 多个不可变引用
    let mut data = 42;
    let r1 = &data;
    let r2 = &data; // OK：多个不可变引用
    println!("r1={}, r2={}", r1, r2);
    // r1 和 r2 不再使用后，可以创建可变引用
    let r3 = &mut data;
    *r3 = 99;
    println!("After mut: {}", data);

    // 5. Box<T> - 堆分配的拥有型指针
    let boxed = Box::new(42);
    println!("Boxed value: {}", boxed); // 自动解引用
    println!("Boxed size on stack: {} bytes", std::mem::size_of_val(&boxed));

    // Box 常用于递归类型
    #[derive(Debug)]
    #[allow(dead_code)]
    enum List {
        Cons(i32, Box<List>),
        Nil,
    }
    let list = List::Cons(1, Box::new(List::Cons(2, Box::new(List::Nil))));
    println!("List: {:?}", list);

    // 6. 切片引用 &[T]（类似 Zig 的 []const T）
    let arr = [10, 20, 30, 40, 50];
    let slice = &arr[1..4]; // [20, 30, 40]
    println!("Slice: {:?}, sum = {}", slice, sum(slice));

    // 7. 原始指针（仅在 unsafe 中解引用）
    let val = 42i32;
    let raw_ptr: *const i32 = &val;
    let raw_mut: *mut i32 = &val as *const i32 as *mut i32;
    unsafe {
        println!("Raw pointer: {}", *raw_ptr);
        // 可以做指针算术
        let arr2 = [100i32, 200, 300];
        let p = arr2.as_ptr();
        println!("arr2[0] = {}", *p);
        println!("arr2[2] = {}", *p.add(2)); // 指针算术
    }
    let _ = raw_mut; // 避免 unused 警告

    // 8. 自动解引用（Deref trait）
    let s = String::from("hello");
    // String 自动解引用为 &str
    fn takes_str(s: &str) {
        println!("Got: {}", s);
    }
    takes_str(&s); // String -> &str 自动转换

    // 9. 字节数据重解释（类似 C 的类型双关）
    let bytes: [u8; 4] = [0x01, 0x00, 0x00, 0x00];
    let int_val: u32 = u32::from_le_bytes(bytes);
    println!("Bytes as u32: {}", int_val);
}
